### PR TITLE
Refactor complete for quickbooks, tasks, users

### PR DESCRIPTION
**Pull request by ruuushhh**

### CODE REVIEW

File: apps/quickbooks_online/actions.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            status=status.HTTP_200_OK
+        )
+    except QBOCredential.DoesNotExist:
+        return Response(
+            data={
+                'message': 'QBO credentials not found in workspace'
+            },
+            status=status.HTTP_400_BAD_REQUEST
+        )
+    except WrongParamsError:
```
Comment (ashwin1111) [inline apps/quickbooks_online/actions.py:32]: Club InvalidTokenError, both does the same job. Adjust the response message in a slightly generic way to cover these 2 exceptions

File: apps/quickbooks_online/actions.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    if workspace.destination_synced_at:
+        time_interval = datetime.now(timezone.utc) - workspace.destination_synced_at
+
+    if workspace.destination_synced_at is None or time_interval.days > 0:
+        quickbooks_credentials = QBOCredential.get_active_qbo_credentials(workspace_id)
+        quickbooks_connector = QBOConnector(quickbooks_credentials, workspace_id=workspace_id)
+        quickbooks_connector.sync_dimensions()
+
+        workspace.destination_synced_at = datetime.now()
+        workspace.save(update_fields=['destination_synced_at'])
```
Comment (ashwin1111) [inline apps/quickbooks_online/actions.py:93]: new line at end of file

File: apps/quickbooks_online/views.py [hunk 3]
```diff
@@ -27,32 +19,23 @@ class VendorView(generics.ListCreateAPIView):
     """
     Vendor view
     """
+    queryset = DestinationAttribute.objects.all()[:10]
```
Comment (ashwin1111) [inline apps/quickbooks_online/views.py:22]: is slicing working properly? noticed a message on slack around this area, pls confirm

File: apps/quickbooks_online/views.py [hunk 4]
```diff
@@ -27,32 +19,23 @@ class VendorView(generics.ListCreateAPIView):
...
-    def get_queryset(self):
-        search_term = self.request.query_params.get('search_term')
-        if search_term:
-            return DestinationAttribute.objects.filter(
-                attribute_type='VENDOR', active=True, workspace_id=self.kwargs['workspace_id'],value__icontains=search_term).order_by('value')[:10]
-        return DestinationAttribute.objects.filter(
-            attribute_type='VENDOR', active=True, workspace_id=self.kwargs['workspace_id']).order_by('value')[:10]
-
+    filter_backends = (DjangoFilterBackend,)
+    filterset_fields = {'value': {'icontains'}, 'attribute_type': {'exact', 'in'}, 'active':{'exact'}, 'workspace_id': {'exact'}}
```
Comment (ashwin1111) [inline apps/quickbooks_online/views.py:26]: `'active': {'exact'}` one space after colon. Also, should we send workspace_id explicitly from client as query param or does it pick from the URL?

File: apps/quickbooks_online/views.py [hunk 5]
```diff
@@ -27,32 +19,23 @@ class VendorView(generics.ListCreateAPIView):
...
+    ordering_fields = ['value']
 
 class EmployeeView(generics.ListCreateAPIView):
     """
     Employee view
     """
+    queryset = DestinationAttribute.objects.all()
```
Comment (ashwin1111) [inline apps/quickbooks_online/views.py:33]: why there is no limit to this API?

File: apps/quickbooks_online/views.py [hunk 6]
```diff
@@ -61,47 +44,8 @@ class PreferencesView(generics.RetrieveAPIView):
...
-                data={
-                    'message': 'Invalid token, try to refresh it'
-                },
-                status=status.HTTP_400_BAD_REQUEST
-            )
-
+        return get_preferences(kwargs['workspace_id'])
+       
 
 class BillPaymentView(generics.CreateAPIView):
```
Comment (ashwin1111) [inline apps/quickbooks_online/views.py:50]: this is a dead endpoint, pls remove

File: apps/quickbooks_online/views.py [hunk 7]
```diff
@@ -186,67 +98,32 @@ class DestinationAttributesView(generics.ListAPIView):
...
-            'active': True
-        }
-
-        if display_name:
-            display_name = display_name.split(',')
-            filters['display_name__in'] = display_name
-
-        return DestinationAttribute.objects.filter(**filters).order_by('value')
-
+#slice problem
```
Comment (ashwin1111) [inline apps/quickbooks_online/views.py:108]: todo

File: apps/quickbooks_online/views.py [hunk 8]
```diff
@@ -186,67 +98,32 @@ class DestinationAttributesView(generics.ListAPIView):
...
-        if search_term:
-            filters['value__icontains'] = search_term
-
-        if active and active.lower() == 'true':
-            filters['active'] = True
-
-        return DestinationAttribute.objects.filter(**filters).order_by('value')[:30]
-
+    filter_backends = (DjangoFilterBackend,)
+    filterset_fields = {'workspace_id': {'exact'}, 'attribute_type': {'exact', 'in'}, 'display_name': {'exact', 'in'}, 'active': {'exact'}}
```
Comment (ashwin1111) [inline apps/quickbooks_online/views.py:117]: why are we not adding value? the existing view would filter by value as well

File: apps/tasks/urls.py [hunk 9]
```diff
@@ -18,6 +18,6 @@
 from .views import TasksView, TasksByIdView
 
 urlpatterns = [
-    path('', TasksByIdView.as_view()),
+    path('<int:id>/', TasksByIdView.as_view()),
```
Comment (ashwin1111) [inline apps/tasks/urls.py:21]: these are dead endpoints, let's remove, please verify once

File: apps/quickbooks_online/views.py (hunk 4 above)
Comment (ruuushhh) [reply apps/quickbooks_online/views.py:26]: no we will have to send explicitly

Comment (ashwin1111) [reply apps/quickbooks_online/views.py]: we should find a way to pick it from url

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**